[Suggestion] Batch textDocument_didChange

HerringtonDarkholme opened the pull request:
Language Server Protocol implies reading (or writing) is single threaded.

So when frequent text change take places, server must read all textDocument_didChanges, even if most of them are stale. The same applies to client writing. 

If we can throttle the frequency, server performance, especially completion, will improve. We need also implement a guard flag to refresh document before sending completion/definition requests.

We can also implement incremental `contentChange` item instead of full text reload, but that's another story.

This pull request just illustrates the idea. Feel free to close it. Also, naming convention is arguable.

HerringtonDarkholme commented on rplugin/python3/LanguageClient/LanguageClient.py:
```diff
@@ -51,6 +51,7 @@ def __init__(self, nvim):
         type(self)._instance = self
         self.serverCommands = self.nvim.eval(
                 "get(g:, 'LanguageClient_serverCommands', {})")
+        self.skip_threshold = 0.5
```
Strangely I cannot call `nvim.eval` multiple times.

autozimu commented:
Thanks for the work.

I do notice there are some stall change notifications, especially in completion, which can be removed by comparing file content with cached file content. PR is very welcome to fix this.

For the debouse enhancement,  as discussed by https://github.com/autozimu/LanguageClient-neovim/issues/29, generally I don't think that's a problem. But I might be wrong. If you observe LanguageClient-neovim taxing too much on language servers, please let me know. Or we can observe and mimic the behaviours in vscode.

I agree that ideally we should send patches instead of full text, while that's quite some work even with the powerful help of python stdlib. https://github.com/neovim/neovim/pull/5269 is interesting, let's see where it goes.

HerringtonDarkholme commented:
> For the debouse enhancement, as discussed by #29, generally I don't think that's a problem

This optimization is mainly for large files. 
One example I encounter is TypeScript's compiler code. A typical compiler file is about 10k lines. TypeScript server is highly optimized and can elegantly handle files with several thousands lines. 
But for compiler code with 20k+ lines it feels obviously slow and jerky. Without NeoVim's support, I find it hard to implement patch like text change now. Yet debounce is fairly straightforward to do.

Yes, I agree that incremental change is more desirable, ideally.

But in reality building a performant language service is herculean task. I have investigated Python's Jedi, Golang' gocode and php language server which power language servers underhood. It seems they all handle whole file text internally. (I might be wrong because I skimmed these code). So passing patch update to server might not achieve desirable performance. For example, Jedi always accept the whole file as argument. So only throttling will get performance boost. Adapting language service to handle incremental parsing / type checking in forseeable future is not possible for most language, IMHO.

autozimu commented:
All right. I see your pain when working with super large files with specific language servers.

One suggestion though, can you make this behind a configuration flag? The delay of sending change notifications might introduce delays for completion. Users using performant language servers shall not pay these overheads.

HerringtonDarkholme commented:
The new commit makes a workaround with neovim's bug reported here https://github.com/neovim/neovim/issues/6353. 

The throttling is now opt-in by setting `g:LanguageClient_changeThreshold`. By default, threshold is 0s, that is, client will not batch any update.

Is this good?

autozimu commented on rplugin/python3/LanguageClient/LanguageClient.py:
```diff
@@ -890,6 +919,8 @@ def textDocument_codeAction(
             cbs: List = None) -> None:
         logger.info("Begin textDocument/codeAction")
+
+        # self.sync_doc(uri)
```
Is this comment intentional?

HerringtonDarkholme replied:
I don't know whether codeAction requires an update. So I just comment here.

autozimu replied:
Yes, it needs files up to date.

autozimu commented on rplugin/python3/LanguageClient/LanguageClient.py:
```diff
@@ -49,8 +49,9 @@ def __init__(self, nvim):
-        self.serverCommands = self.nvim.eval(
+        self.serverCommands = nvim.eval(
                 "get(g:, 'LanguageClient_serverCommands', {})")
+        self.skip_threshold = 0
```
I'd prefer variables sync with vim side instead of inventing another name. What about change it to `changeThreshold`?

HerringtonDarkholme replied:
I will change it.

autozimu commented:
Overall looks good to me.

I don't know exactly how hard it would be to add tests. If it is not easy to achieve, you can skip it over or left it to me.

And please add docs.

Thanks

HerringtonDarkholme commented:
I have added document and addressed feedback. Sadly I didn't manage to build test environment locally. So I skipped the test part.

Additionally, it seems hard to test throttling behavior since it needs client's internal information. The current test is black box, though.

autozimu commented:
That's good enough. Thank you! I will merge it in shortly.

For the test part, I will try add one or two some time next week.



On Sun, Mar 26, 2017 at 04:33 Herrington Darkholme <notifications@github.com>
wrote:

> I have added document and addressed feedback. Sadly I didn't manage to
> build test environment locally. So I skipped the test part.
>
> Additionally, it seems hard to test throttling behavior since it needs
> client's internal information. The current test is black box, though.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/autozimu/LanguageClient-neovim/pull/34#issuecomment-289266335>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYt7w0olQIqT8BoqyIWIKNjS3ksYh8gks5rpiLrgaJpZM4MmneI>
> .
>
